v0.11 (#6)
boa-dev/boa-dev.github.io: _posts/2021-01-10-boa-release-11.md

---
layout: post
title: "Boa release v0.11"
author: Boa Developers
---

Boa has reached a new release. v0.11, our biggest one yet. Since v0.10 we've closed 77 issues, merged 129 pull requests. The engine has been faster and more compliant to the spec. Below are some of the highlights but please see the [changelog link] for more information.

> ## Test 262

Reviewer: In this section we could add an image of the progress, like the one in https://github.com/boa-dev/boa/pull/1065.

Author: Do you have a light and dark mode version of that picture?

Reviewer: We have no dark mode, but I will create the pictures and send them to you via Discord :)

Author: Thanks, if the background could match the color of https://boa-dev.github.io/ that would be nice

> Test262 is the implementation conformance test suite maintained by TC39. It's used by nearly all engines to measure how conformant they are to the specification. Boa pulls the tests in-tree and runs against PRs and on a nightly basis. Test262 doesn't sit still, so since the last release we've updated our reference to make sure we're running against the latest spec. You can find more about Test262 [here](https://github.com/tc39/test262).

Reviewer 2: I don't think the term nightly fits here, what do you think?
```suggestion
Test262 is the implementation conformance test suite maintained by TC39. It's used by nearly all engines to measure how conformant they are to the specification. Boa pulls the tests in-tree and runs them against all PRs. Test262 doesn't sit still, so since the last release we've updated our reference to make sure we're running against the latest spec. You can find more about Test262 [here](https://github.com/tc39/test262).
```

> Since v0.10 we have almost doubled on spec conformance, and reduced panics from hundreds to less than 50. A year ago Boa didn't even track itself against Test262 so it was difficult to know how compliant we were to the spec, today not only do we track all changes against Test262 but we can see progress on a PR to PR basis.

Reviewer: ```suggestion
Since v0.10 we have almost doubled on spec conformance, and reduced panics from hundreds to 25 (tracked in [#817](https://github.com/boa-dev/boa/issues/817)). A year ago Boa didn't even track itself against Test262 so it was difficult to know how compliant we were to the spec, today not only do we track all changes against Test262 but we can see progress on a PR to PR basis.
```

This was worked on by @jevancc and his [changes](https://github.com/boa-dev/boa/pull/915/files) have improved performance overall.

## Embedding examples

We are still working on what the public API should look like, some of these decisions are driven by feedback and the constant changing way which Boa works.  
Not only you can run Boa against javascript today you can also embed your own objects and classes into the engine before it's ran.

We've offered an [example](https://github.com/boa-dev/boa/blob/master/boa/examples/classes.rs) to show how a class can be constructed in Rust then added to the environment before executing your scripts.
This should offer a great insight into how you can interop Rust with JavaScript by using Boa.

> If there are any examples you would like added, please raise an issue on the main repository.

Reviewer: We are missing some examples until https://github.com/boa-dev/boa/pull/634 (or an equivalent PR) get merged. Should we add some code examples here? Or at least mention that using a `Context` we can register global functions, global classes and so on.

We could mention the `Context::register_global_function()`, `Context::register_global_class()` and `Context::register_global_property()` functions, to make our users understand how to extend Boa.

Author: Yeah I like the sound of that

## Generating bytecode

Today Boa walks the tree of the AST, although easy to implement it's not ideal for performance.  
We are loooking to do code generation which can then be interpreted. This gradual process may happen over many releases until at some point we can switch implementations under the hood. Our steps are:

- ~~Experiment with VM Path on isolated branch~~
- ~~Experiment with generating more simpler instructions~~
- Build up parity with current implementation
- Run test suite over VM path including conformance tests
- Switch over once performance is steady

> - Optimize code generation and intreperter

Reviewer 2: Is it possible to use a checklist like on Github? i.e.

- [x] one
- [ ] two

I think it has better readability than strikethrough.

Reviewer: It will work in GitHub, but not sure if in GitHub pages.

Author: yeah, sadly it doesn’t work :(

## Thank You

> Everything in this release has been such a huge effort, we thank everyone who contributed, whether it was features, fixes or raising bugs.

Reviewer: Here I would mention the [contributors](https://github.com/boa-dev/boa/graphs/contributors?from=2020-10-02&to=2021-01-12&type=c) since the last release. Especially for the ones outside of the org, and the new memberships in the organization.

I would also add links to the [good first issues](https://github.com/boa-dev/boa/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22) and maybe also [issues where help is wanted](https://github.com/boa-dev/boa/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22), so that potential contributors know where to start.
